// base/big-record.js

var db = require('base/dbmanager');
var moment = require('moment');

module.exports = BigRecord;

function BigRecord() {
    this.id = 0;//数据库对应id
    this.roomId = 0;//房间号
    this.setting1 = "";//设置
    this.setting2 = "";
    this.setting3 = "";
    this.setting4 = "";
    this.setting5 = "";
    this.createUserId = 0;//创建者
    this.users = "";//用户
    this.createAt = null;
    this.updateAt = null;
    this.scores = "";//得分信息 对应users的顺序
    this.playersInfo = "";

    this.save = function (callback) {
        var cTime = moment().format('YYYY-MM-DD HH:mm:ss');
        if (this.createAt == null) {
            this.createAt = cTime;
        }
        this.updateAt = cTime;
        /*default create*/
        var sql = "INSERT INTO " +
            "big_records(roomId,setting1,setting2,setting3,setting4,setting5,createUserId,users,createAt,updateAt,scores,playersInfo) " +
            "VALUES (?,?,?,?,?,?,?,?,?,?,?,?)";
        var sqlParam = [this.roomId,this.setting1,this.setting2,JSON.stringify(this.setting3),this.setting4,this.setting5,this.createUserId,JSON.stringify(this.users),this.createAt,this.updateAt,JSON.stringify(this.scores),this.playersInfo];
        /*update*/
        if (this.id > 0) {
            sql = "UPDATE big_records SET scores = ?,updateAt = ?,playersInfo = ? WHERE id = ?";
            sqlParam = [JSON.stringify(this.scores),this.updateAt,this.playersInfo,this.id];
        }
        /*保存 或更新数据*/
        var self = this;
        db.query(sql,sqlParam,function (error,result) {
            if (self.id == 0) {
                self.id = result['insertId'];
                //console.log(self.id);
            }
            console.log('big-record save success! id = ' + self.id);
            callback(error,result);
        });
    };
}
BigRecord.getRecordfromId = function(id,callback){
    var sql = "SELECT * FROM big_records WHERE id = ?";
    var sqlParam = [id];
    db.query(sql,sqlParam,function (error,result) {
        if(!error){
            callback(result);
        }

    });
}
BigRecord.getRecords = function (userId,callback) {
    var sql = "SELECT * FROM big_records WHERE users LIKE \"%?%\" ORDER BY updateAt DESC LIMIT 10";
    var sqlParam = [userId];
    db.query(sql,sqlParam,function (error,result) {
        for (var i = 0;i < result.length;i++) {
            result[i]['createAt'] = moment(result[i]['createAt']).format("YYYY-MM-DD HH:mm:ss");
            result[i]['updateAt'] = moment(result[i]['updateAt']).format("YYYY-MM-DD HH:mm:ss");
            result[i]['userid'] =userId;
        }
        //console.log(result,"--------------------??62");
        callback(error,result);
    });
};
