// base/record.js

var db = require('base/dbmanager');
var moment = require('moment');

module.exports = Record;

/*
    一个小局
    存储一个小局
*/
function Record() {

    this.parentId = 0;//总局id
    this.userId = 0;//用户
    this.gameIndex = 0;//第几局
    this.score = 0;//得分

    this.info = "";
    this.info0 = "";
    this.info1 = "";
    this.info2 = "";
    this.info3 = "";
    this.info4 = "";
    this.info5 = "";
    this.info6 = "";
    this.info7 = "";
    this.info8 = "";//8 9 可以存储大数据
    this.info9 = "";
    this.creatAt = "";

    this.save = function (parentId,callback) {
        console.log(parentId);
        if (parentId == 0 || parentId == null) return;
        var cTime = moment().format('YYYY-MM-DD HH:mm:ss');
        var sql = "INSERT INTO " + "records(info,info0,info1,info2,info3,info4,info5,info6,info7,info8,info9,creatAt,parentId,userId,gameIndex,score) " + "VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)";
        var sqlParam = [
        this.info, this.info0, this.info1, this.info2, this.info3,
        this.info4, this.info5, this.info6, this.info7, this.info8,
        this.info9,  cTime,      parentId,   this.userId, this.gameIndex,
        this.score];
        db.query(sql,sqlParam,function (error,result) {
            if (error) {
                console.log( sqlParam );
                console.log( sql );
                console.log( error,"---------------------records--41" );
                return;
            }
            callback(error,result);
            //console.log('record save success!');
        });
    };

    this.getRecords = function(parentId,userId,callback) {
        if (parentId == 0 || parentId == null || userId == 0 || userId == null) {
            callback([]);
            return;
        }
        var sql = "SELECT * FROM records WHERE parentId = ? AND userId = ?";
        var sqlParam = [parentId,userId];
        db.query(sql,sqlParam,function (error,result) {
            if (error) {
                callback([]);
                return;
            }
            for (var i = 0;i < result.length;i++) {
                result[i]['creatAt'] = moment(result[i]['creatAt']).format("MM-DD HH:mm")
            }
            callback(result);
        });
    };
}

Record.getHallInfo = function(callback) {
    var sql = "SELECT * FROM hall";
    db.query(sql,function(error,result) {
        if (error) {
            callback({});
            return;
        }
        if (result.length > 0) {
            callback(result[0]);
        }
        else {
            callback({});
        }
    });
};
